std/variant.d

```diff
@@ -102,6 +102,23 @@ struct This;
 
 private alias This2Variant(V, T...) = AliasSeq!(ReplaceType!(This, V, T));
 
+// copied from std.traits and modified the base case
```

Comment:
You could reduce the code duplication by simply checking for an empty tuple (and returning the pointer-alignment then, which seems like a valid guess, although `max(size_t.alignof, real.alignof)` would be safer), and otherwise use the `std.traits` version.

Reply:
Made the `std.traits` version `package`, so I can reuse it like you said.

As for setting the maximum alignment of `real` or `size_t`, that hits the Android/x64 segfaults I mentioned in the ldc pull, when using `align(16)` for the union. It's using 128-bit alignment then, as the `real` alignment is always larger than 64-bit `size_t` there, and then segfaults when trying to `MOVAPS` a 64-bit aligned location into the FP register `$xmm0`. With `size_t` alignment, it uses a `MOVUPS` instead.

Another case where LLVM isn't expecting that `__float128` alignment?

Reply:
> and then segfaults when trying to MOVAPS a 64-bit aligned location into the FP register $xmm0

That's exactly what shouldn't happen when aligning the union at 16 bytes. What that does though is change the min size of the union to 16 bytes, which might be unexpected for some test.

I mentioned `max(size_t.alignof, real.alignof)` for an empty `AllowedTypes` tuples (=> arbitrary types) as that should make sure it can be used as `real` too on platforms like Android x64. I'm pretty sure you'll hit alignment issues otherwise, just like the ones you currently get when using that max, for some bizarre reason.

Reply:
So we're clear, [the original failing test](https://github.com/ldc-developers/phobos/blob/fd5facfe0456d8f7ae16af8415f1ea01606c917f/std/variant.d#L1317) passes now: it's [the other tests that start to fail](https://github.com/ldc-developers/phobos/blob/fd5facfe0456d8f7ae16af8415f1ea01606c917f/std/variant.d#L39) if I use `real.alignof` when no types are passed to `Variant`, as [I mentioned before](https://github.com/ldc-developers/ldc/pull/2917#issuecomment-445751275). Looking into it more in a debugger and after hacking that first test block down to just two lines, `Variant b = 42; assert(b.type == typeid(int));`, it loads a 64-bit-aligned `_D3std7variant__T8VariantNVmi32ZQp6__initZ` and tries to `MOVAPS` the first 128-bit word and fails.

If I use this pull with `size_t.alignof` instead, all the `std.variant` tests pass, and it loads the FP registers from that same 64-bit aligned initializer with `MOVUPS` instructions instead.

I honestly don't care how this stuff is aligned, as long as the tests pass. :wink:

Reply:
> it loads a 64-bit-aligned _D3std7variant__T8VariantNVmi32ZQp6__initZ

That seems to be the issue then - our (LDC) init symbols seem to violate the alignment.

Reply:
Well that would have been quite surprising, so I checked - and it looks fine: https://run.dlang.io/is/q45ykS

```LL
@onlineapp.VariantN.__init = constant %onlineapp.VariantN zeroinitializer, comdat, align 16
```

So I guess it could be a linker problem.

Reply:
Sorry, I wasn't really focused - `movups` is of course used for `align(8)`. `movaps` should be fine, and the reduced testcase suggests LDC does the right thing by requesting 16-bytes alignment for the init symbol -> I guess it's really a linker issue, wouldn't be the first of this kind.

Reply:
Alright, trying it.

Reply:
Works on the auto-tester, could the Android/x64 linker be why `std.variant` passed its tests when I used `align(16)` and linked that module alone, but failed when I linked it into the test runner? If you think it's more correct this way, fine with me.

Reply:
Ah, very nice then.